Created CSourceMetaData module for model metadata

manupa-arm opened the pull request:
* Currently, there is a MetaData module to capture constants
  conditionaly if the runtime modules implement const init
  PackedFuncs. However, this one relies on a load process
  in which the metadata is created on volatile memory that
  may be not usable in uTVM environments.
* There is a need for model level metadata that is valid
  across all runtime modules such as the func registry
  when creating a system-lib.
* This commit implements a CSoureMetaData module to hold
  func registry that collects function names from the
  runtime module and generates a c source file to be
  linked with final artifact.
* Modified and added export_library for utvm

Discuss : https://discuss.tvm.apache.org/t/csourcemetadata-module-a-csourcemodule-to-hold-function-registry-for-utvm/8554

manupa-arm commented:
@areusch @tqchen @mbaret @zhiics
Please take a look when you have a minute :).

zhiics commented on src/relay/backend/build_module.cc:
```diff
@@ -493,18 +493,14 @@ class RelayBuildModule : public runtime::ModuleNode {
-    Array<tvm::runtime::Module> ext_mods = graph_codegen_->GetExternalModules();
-    // TODO(zhiics) We should be able to completely switch to MetadataModule no
-    // matter whether there are external modules or not.
-    if (!ext_mods.empty()) {
```
I think this would duplicate the param in the .so and .params file. I had a plan to handle it but didn't have cycles.

zhiics commented on src/relay/backend/contrib/codegen_c/codegen.cc:
```diff
@@ -215,20 +215,24 @@ class CodegenC : public MemoizedExprTranslator<std::vector<Output>>, public Code
     auto out = builder.VisitExpr(func->body);
-    code_stream_ << builder.JIT(out);
-
-    return {sid, builder.const_vars_};
```
I think sid is the same to ext_func_id_.

manupa-arm replied:
Removed sid and just used ext_func_id_

zhiics commented on src/relay/backend/contrib/codegen_c/codegen.cc:
```diff
@@ -259,18 +263,13 @@ class CSourceCodegen : public CSourceModuleCodegenBase {
     const auto* pf = runtime::Registry::Get("runtime.CSourceModuleCreate");
     ICHECK(pf != nullptr) << "Cannot find csource module to create the external runtime module";
-    return (*pf)(code, "c", sym, variables);
+    return (*pf)(code, "c", Array<String>{func_name}, sym, variables);
```
I think we can use sym directly. It should be the same to func_name

manupa-arm replied:
Used func_name as it is more clearer term and removed sym.

zhiics commented on src/relay/backend/contrib/dnnl/codegen.cc:
```diff
@@ -413,7 +413,8 @@ class DNNLModuleCodegen : public CSourceModuleCodegenBase {
     ICHECK(pf != nullptr) << "Cannot find csource module to create the external runtime module";
-    return (*pf)(code, "c", sym, variables);
+    // TODO(@manupa-arm): pass the function names to enable system-lib creation
+    return (*pf)(code, "c", Array<String>{}, sym, variables);
```
Do we need an array? it seems that a csourcemodule is only one function though.

manupa-arm replied:
When we target "c", in it can have more than one function which also uses the same CSourceModuleCreate interface. See codegen_c_host.cc

zhiics commented on src/relay/backend/vm/compiler.cc:
```diff
@@ -1145,11 +1145,9 @@ void VMCompiler::Codegen() {
-    exec_->lib = codegen::CSourceModuleCreate(";", "");
-  }
-  if (!ext_mods.empty()) {
-    exec_->lib = codegen::CreateMetadataModule(params_, exec_->lib, ext_mods);
```
same here

zhiics reviewed (commented):
half way done. will review the module part later.

areusch reviewed (changes requested):
did an initial pass

manupa-arm commented on python/tvm/micro/build.py:
```diff
@@ -74,8 +74,8 @@ def path(self):
     "cflags": ["-std=c11"] + _COMMON_CFLAGS,
-    "ccflags": ["-std=c++11"] + _COMMON_CFLAGS,
-    "ldflags": ["-std=c++11"],
+    "ccflags": ["-std=gnu++11"] + _COMMON_CFLAGS,
```
I think current implementation of linked-params uses hexadecimal notation that requires gnu++. This is partly because export_library saves c-source modules .cc.

manupa-arm replied:
Alright, I think the correct thing to do is make export library save the runtime modules as .c not .cc. Thus, did that and reverted these changes

areusch replied:
ah, I just realized that was what was going on here. I agree we shouldn't be generating c++ for the c backend.

manupa-arm commented on python/tvm/runtime/module.py:
```diff
@@ -268,6 +268,11 @@ def export_library(self, file_name, fcompile=None, addons=None, **kwargs):
+
         kwargs : dict, optional
             Additional arguments passed to fcompile
         """
```
Done

manupa-arm commented on python/tvm/runtime/module.py:
```diff
@@ -330,25 +341,25 @@ def export_library(self, file_name, fcompile=None, addons=None, **kwargs):
                 files.append(path_cc)
 
-        if has_c_module:
+        if has_c_module and not file_name.endswith(".tar"):
```
Done

manupa-arm commented on src/target/source/source_module.cc:
```diff
@@ -46,20 +48,25 @@ using runtime::SaveBinaryToFile;
- * \param modules The submodules that will be wrapped, e.g. CSource modules that
- *        contain vendor library calls or customized runtime modules.
- *
+ * \param modules All the modules that needs to be imported inside the metadata module(s).
```
Done

manupa-arm commented on python/tvm/runtime/module.py:
```diff
@@ -268,6 +268,11 @@ def export_library(self, file_name, fcompile=None, addons=None, **kwargs):
             to that format. Otherwise, will use default format "o".
 
+        workspace_dir : str, optional
+            The name of the workspace dir to create intermediary
```
Done

manupa-arm commented:
@tqchen @areusch @zhiics , I ve addressed all the comments and aligned with what we discussed in the RFC.
Can you have another look when you have time? If its OK, please approve explicitly. Thanks!

areusch commented on src/target/source/source_module.cc:
```diff
@@ -46,40 +48,68 @@ using runtime::SaveBinaryToFile;
+
+    // TODO(@manupa-arm) : we should be able to use csource_metadata
+    // if the variables are empty when all the runtime modules implement get_func_names
+    if (arrays.empty() && DSOExportable(mod) && target->kind->name == "c") {
```
perhaps it is better to enable this based on --link-params?

manupa-arm replied:
Not sure I follow; 

Irrespective of link-params, we would need the DSOModules to fall under c-source metadata module if the target is "c" and the runtime module does not require constant loading.

areusch replied:
oh, sorry you're right.

areusch reviewed (changes requested):
@manupa-arm looks like we are converging now. just a couple more things I want to resolve

manupa-arm commented on python/tvm/micro/build.py:
```diff
@@ -74,8 +74,8 @@ def path(self):
-    "ccflags": ["-std=c++11"] + _COMMON_CFLAGS,
-    "ldflags": ["-std=c++11"],
+    "ccflags": ["-std=gnu++11"] + _COMMON_CFLAGS,
+    "ldflags": ["-std=gnu++11"],
```
You may be right; Let me try this one.

manupa-arm replied:
Reverted them back to c++11 and made export library save the runtime modules to .c instead .cc

zhiics commented on src/target/source/source_module.cc:
```diff
@@ -46,40 +48,68 @@ using runtime::SaveBinaryToFile;
-  init_m.Import(dso_module);
-  for (const auto& it : modules) {
-    init_m.Import(it);
+  if (target.defined()) {
```
this can if `arget.defined()` && `target->kind->name`

manupa-arm replied:
Done

zhiics commented on src/target/source/source_module.cc:
```diff
@@ -46,40 +48,68 @@ using runtime::SaveBinaryToFile;
+    // TODO(@manupa-arm) : we should be able to use csource_metadata
+    // if the variables are empty when all the runtime modules implement get_func_names
+    if (arrays.empty() && DSOExportable(mod) && target->kind->name == "c") {
+      csource_modules.push_back(mod);
```
why do we need to extract csource_module? Originally the target module is a DSOModule. The other external module are parallel to it like the following:

```
MetadataModule
    -- DSO
    -- C source1
    -- C source2
    -- other type of modules
```
Each sub module manages their own constants. Now we combine c source modules to the DSO. What's the advantage of it?

areusch replied:
with C runtime on bare metal, we have no `dlsym`, so we must assemble a string-to-function pointer lookup ourselves. that's why we pass them to metadata module here

manupa-arm replied:
Yes, that is correct.

We would only need the current MetadataModule (which is non-DSOExportable) if the runtime module's requires the constants to be copied in the init process -- which is something we dont want in bare-metal. In-fact if the non of external modules have constants extracted (e.g., ethos-n and ethos-u will be managing their own constants), we dont even want MetadataModule (which is non-DSOExportable) get imported to the module hierarchy.

Moreover, the purpose of the c-source metadata module is to provide an global entity for all DSOExportable modules a place to hold global data of the model. Initially its the function registry which @areusch describes here.

manupa-arm commented:
Ping @areusch @zhiics.
Let me know if you have any more concerns, I ll be disappearing for holidays from tomorrow. 
Thus, would like to address if there is anything today. Thanks!

areusch reviewed (approved):
I think all my concerns are addressed. @zhiics anything further?

zhiics commented:
Thanks @manupa-arm @areusch
